// @substrate/connect-known-chains/dist/index.js
import {
  chainSpec as chainSpec14
} from "./chunk-5AT6HN4B.js";
import {
  chainSpec as chainSpec13
} from "./chunk-CNQZBKNT.js";
import {
  chainSpec as chainSpec4
} from "./chunk-SA2EJ2JV.js";
import {
  chainSpec as chainSpec7
} from "./chunk-SUER3VNA.js";
import {
  chainSpec as chainSpec11
} from "./chunk-KTJHKCJG.js";
import {
  chainSpec as chainSpec3
} from "./chunk-NJC3XV76.js";
import {
  chainSpec as chainSpec8
} from "./chunk-EJVEB6K2.js";
import {
  chainSpec as chainSpec12
} from "./chunk-UCCKYOP4.js";
import {
  chainSpec as chainSpec2
} from "./chunk-U24VUTNT.js";
import {
  chainSpec as chainSpec6
} from "./chunk-OUWB7QGP.js";
import {
  chainSpec as chainSpec10
} from "./chunk-JCIYR5ZS.js";
import {
  chainSpec as chainSpec15
} from "./chunk-4OR5S2TK.js";
import {
  chainSpec
} from "./chunk-4KTZNILW.js";
import {
  chainSpec as chainSpec5
} from "./chunk-2EEM5CDM.js";
import {
  chainSpec as chainSpec9
} from "./chunk-CM7MO2UQ.js";
export {
  chainSpec2 as ksmcc3,
  chainSpec6 as ksmcc3_asset_hub,
  chainSpec10 as ksmcc3_bridge_hub,
  chainSpec15 as paseo,
  chainSpec as polkadot,
  chainSpec5 as polkadot_asset_hub,
  chainSpec9 as polkadot_bridge_hub,
  chainSpec13 as polkadot_collectives,
  chainSpec4 as rococo_v2_2,
  chainSpec7 as rococo_v2_2_asset_hub,
  chainSpec11 as rococo_v2_2_bridge_hub,
  chainSpec3 as westend2,
  chainSpec8 as westend2_asset_hub,
  chainSpec12 as westend2_bridge_hub,
  chainSpec14 as westend2_collectives
};
