// @polkadot/util-crypto/cjs/mnemonic/wordlists/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.chineseTraditional = exports.chineseSimplified = exports.korean = exports.japanese = exports.italian = exports.french = exports.spanish = exports.english = void 0;
var en_js_1 = require("./en.js");
Object.defineProperty(exports, "english", { enumerable: true, get: function () { return __importDefault(en_js_1).default; } });
var es_js_1 = require("./es.js");
Object.defineProperty(exports, "spanish", { enumerable: true, get: function () { return __importDefault(es_js_1).default; } });
var fr_js_1 = require("./fr.js");
Object.defineProperty(exports, "french", { enumerable: true, get: function () { return __importDefault(fr_js_1).default; } });
var it_js_1 = require("./it.js");
Object.defineProperty(exports, "italian", { enumerable: true, get: function () { return __importDefault(it_js_1).default; } });
var jp_js_1 = require("./jp.js");
Object.defineProperty(exports, "japanese", { enumerable: true, get: function () { return __importDefault(jp_js_1).default; } });
var ko_js_1 = require("./ko.js");
Object.defineProperty(exports, "korean", { enumerable: true, get: function () { return __importDefault(ko_js_1).default; } });
var zh_s_js_1 = require("./zh-s.js");
Object.defineProperty(exports, "chineseSimplified", { enumerable: true, get: function () { return __importDefault(zh_s_js_1).default; } });
var zh_t_js_1 = require("./zh-t.js");
Object.defineProperty(exports, "chineseTraditional", { enumerable: true, get: function () { return __importDefault(zh_t_js_1).default; } });
